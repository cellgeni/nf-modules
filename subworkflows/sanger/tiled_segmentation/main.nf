include { IMAGING_CELLPOSE as CELLPOSE } from '../../../modules/sanger/imaging/cellpose/main'
include { IMAGING_STARDIST as STARDIST } from '../../../modules/sanger/imaging/stardist/main'
include { IMAGING_INSTANSEG as INSTANSEG } from '../../../modules/sanger/imaging/instanseg/main'
include { IMAGING_DEEPCELL as DEEPCELL } from '../../../modules/sanger/imaging/deepcell/main'
include { IMAGING_MERGEOUTLINES as MERGEOUTLINES } from '../../../modules/sanger/imaging/mergeoutlines/main'
include { IMAGING_GENERATETILES as GENERATE_TILE_COORDS } from '../../../modules/sanger/imaging/generatetiles/main'


workflow TILED_SEGMENTATION {
    take:
    images
    method

    main:
    ch_versions = Channel.empty()
    GENERATE_TILE_COORDS(images)
    ch_versions = ch_versions.mix(GENERATE_TILE_COORDS.out.versions.first())

    images_tiles = GENERATE_TILE_COORDS.out.tile_coords
        .splitCsv(header: true, sep: ",")
        .map { meta, coords ->
            [meta, coords.X_MIN, coords.Y_MIN, coords.X_MAX, coords.Y_MAX]
        }
    tiles_and_images = images_tiles.combine(images, by: 0)
    if (method == "CELLPOSE") {
        CELLPOSE(tiles_and_images.combine(channel.from(params.cell_diameters)))
        wkts = CELLPOSE.out.wkts.groupTuple(by: 0)
        ch_versions = ch_versions.mix(CELLPOSE.out.versions.first())
    }
    else if (method == "STARDIST") {
        STARDIST(tiles_and_images)
        wkts = STARDIST.out.wkts.groupTuple(by: 0)
        ch_versions = ch_versions.mix(STARDIST.out.versions.first())
    }
    else if (method == "INSTANSEG") {
        INSTANSEG(tiles_and_images)
        wkts = INSTANSEG.out.wkts.groupTuple(by: 0)
        ch_versions = ch_versions.mix(INSTANSEG.out.versions.first())
    }
    else if (method == "DEEPCELL") {
        DEEPCELL(tiles_and_images)
        wkts = DEEPCELL.out.wkts.groupTuple(by: 0)
        ch_versions = ch_versions.mix(DEEPCELL.out.versions.first())
    }
    else {
        error("Invalid segmentation method: ${method}. Expected one of: CELLPOSE, STARDIST, INSTANSEG, DEEPCELL")
    }
    MERGEOUTLINES(wkts)
    ch_versions = ch_versions.mix(MERGEOUTLINES.out.versions.first())

    emit:
    geojson = MERGEOUTLINES.out.multipoly_geojsons // channel: [ val(meta), [ geojson ] ]
    versions = ch_versions // channel: [ versions.yml ]
}
