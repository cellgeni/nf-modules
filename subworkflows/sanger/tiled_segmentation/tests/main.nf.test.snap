{
    "nuclear - tif - instanseg": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,9fe55ab05e13024faee6daa2ddcf8e6d"
                    ]
                ],
                "1": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4ca6951145ae8e8a22f95081060586e9",
                    "versions.yml:md5,74ae9680e8565f16244a79acbe2ef1e4"
                ],
                "geojson": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,9fe55ab05e13024faee6daa2ddcf8e6d"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4ca6951145ae8e8a22f95081060586e9",
                    "versions.yml:md5,74ae9680e8565f16244a79acbe2ef1e4"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-03-19T09:25:50.356779643"
    },
    "nuclear - tif - cellpose": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,0d1072acebefc19c48bbcdb17ac57d5f"
                    ]
                ],
                "1": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4ca6951145ae8e8a22f95081060586e9",
                    "versions.yml:md5,7e24ffbd0425ebf6bf537e55ac367001"
                ],
                "geojson": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,0d1072acebefc19c48bbcdb17ac57d5f"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4ca6951145ae8e8a22f95081060586e9",
                    "versions.yml:md5,7e24ffbd0425ebf6bf537e55ac367001"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-03-19T09:24:50.844367069"
    },
    "nuclear - tif - deepcell": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,a3f7237d662627863a30cff6128b989c"
                    ]
                ],
                "1": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4ca6951145ae8e8a22f95081060586e9",
                    "versions.yml:md5,e693a7e9c66cd7a37d9829cae2fc3648"
                ],
                "geojson": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,a3f7237d662627863a30cff6128b989c"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4ca6951145ae8e8a22f95081060586e9",
                    "versions.yml:md5,e693a7e9c66cd7a37d9829cae2fc3648"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-03-19T09:27:14.541410552"
    },
    "nuclear - tif - stardist": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,590cf1b456fbae16631254f0f79121fc"
                    ]
                ],
                "1": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4ca6951145ae8e8a22f95081060586e9",
                    "versions.yml:md5,e3945a4be7e4f9daa0ab9c717a9cce10"
                ],
                "geojson": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,590cf1b456fbae16631254f0f79121fc"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4ca6951145ae8e8a22f95081060586e9",
                    "versions.yml:md5,e3945a4be7e4f9daa0ab9c717a9cce10"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-03-19T09:25:19.772074638"
    }
}